File: Deduction/Parsing/SymbolRegistry.cs
using System;
using System.Collections.Generic;
using Deduction.Abstraction.Connectives;
using Deduction.Abstraction.Constants;

namespace Deduction.Parsing
{
    public class SymbolRegistry
    {
        protected static SymbolRegistry instance = null;
        protected readonly Dictionary<char, Type> connectives;
        protected readonly Dictionary<char, Type> constants;

        public Dictionary<char, Type> Connectives
        {
            get
            {
                return this.connectives;
            }
        }

        public Dictionary<char, Type> Constants
        {
            get
            {
                return this.constants;
            }
        }

        public static SymbolRegistry Instance
        {
            get
            {
                if (SymbolRegistry.instance == null)
                {
                    SymbolRegistry.instance = new SymbolRegistry();
                }

                return SymbolRegistry.instance;
            }
        }

        protected SymbolRegistry()
        {
            this.connectives = new Dictionary<char, Type>()
            {
                { '&', typeof(And) },
                { '|', typeof(Or) },
                { '!', typeof(Not) }
            };

            this.constants = new Dictionary<char, Type>()
            {
                { 't', typeof(True) },
                { 'f', typeof(False) }
            };
        }

        public static char? GetConnectiveSymbol(Type type)
        {
            foreach (KeyValuePair<char, Type> pair in SymbolRegistry.Instance.Connectives)
            {
                if (pair.Value == type)
                {
                    return pair.Key;
                }
            }

            return null;
        }

        public static char? GetConstantSymbol(Type type)
        {
            foreach (KeyValuePair<char, Type> pair in SymbolRegistry.Instance.Constants)
            {
                if (pair.Value == type)
                {
                    return pair.Key;
                }
            }

            return null;
        }
    }
}

File: DeductionCpp/Parsing/SymbolRegistry.cpp
#include "SymbolRegistry.h"

using namespace std;

namespace DeductionCpp { namespace Parsing {
    SymbolRegistry* SymbolRegistry::SymbolRegistryInstance = 0;

    SymbolRegistry::SymbolRegistry()
    {
        SymbolRegistry::SymbolRegistryInstance = this;

        this->Connectives = std::map<char, DeductionCpp::Abstraction::PropositionMemberTypes>();
        this->Constants = std::map<char, DeductionCpp::Abstraction::PropositionMemberTypes>();
    }

    SymbolRegistry& SymbolRegistry::Instance() {
        if (!SymbolRegistry::SymbolRegistryInstance)
        {
            new SymbolRegistry();
            atexit(&SymbolRegistry::Dispose);
        }

        return *SymbolRegistry::SymbolRegistryInstance;
    }

    void SymbolRegistry::Dispose()
    {
        delete SymbolRegistry::SymbolRegistryInstance;
        SymbolRegistry::SymbolRegistryInstance = 0;
    }

    char SymbolRegistry::GetConnectiveSymbol(DeductionCpp::Abstraction::PropositionMemberTypes type)
    {
        SymbolRegistry& instance = SymbolRegistry::Instance();

        for (auto it = instance.Connectives.begin(); it != instance.Connectives.end(); it++)
        {
            if (it->second == type)
            {
                return it->first;
            }
        }

        return NULL;
    }

    char SymbolRegistry::GetConstantSymbol(DeductionCpp::Abstraction::PropositionMemberTypes type)
    {
        SymbolRegistry& instance = SymbolRegistry::Instance();

        for (auto it = instance.Constants.begin(); it != instance.Constants.end(); it++)
        {
            if (it->second == type)
            {
                return it->first;
            }
        }

        return NULL;
    }
} }